### ISSUE TITLE
DeepPartial doesn't work properly if interface has a field named "filter" 

### ISSUE THREAD

**Issue by vasilii-kovalev** [microsoft/TypeScript#39299]
**TypeScript Version:**  3.9.2, 3.7.5

**Search Terms:**
* filter partial
* filter
* filter word

**Code**

```ts
// I've actually tried any other implementations of this generic type from the Internet - the same result.
type DeepPartial<T> = T extends object ? { [K in keyof T]?: DeepPartial<T[K]> } : T;

interface TestInterfaceWithFilter {
  objectWithFilter: {
    filter: object;
  };
  plainObject: object;
}

interface TestInterfaceWithFilter1 {
  objectWithFilter: {
    filter1: object;
  };
  plainObject: object;
}

const testWithFilter: DeepPartial<TestInterfaceWithFilter> = {
  // No errors
  objectWithFilter: [],
  plainObject: [],
};


const testWithFilter1: DeepPartial<TestInterfaceWithFilter1> = {
  // Type 'never[]' has no properties in common with type 'DeepPartial<{ filter1: object; }>'. (2559)
  objectWithFilter: [],
  plainObject: [],
};
```

**Expected behavior:**
* TypeScript complains on `objectWithFilter`'s wrong type in both cases.
* Secondary: TypeScript complains on `plainObject`'s wrong type.

**Actual behavior:**
* TypeScript complains on `objectWithFilter`'s wrong type only if the field with "object" type is not named as "filter".
* TypeScript doesn't complain on `plainObject`'s wrong type.

**Playground Link:** 
[Link](https://www.typescriptlang.org/play?#code/PTAEEkHIDcFNQIYGMAuBXBAbTBPUKAnAS1gBNEA7PAexQAtYDQiBbAB01hdgpQRSLUKAZ1DUAZvjpFRAcx6MiSfDjbxxBaiynxwvRhVgpQAWh2hhCbqAKxhaTCgB0AKBSr4AEViw2ABQQCASwAHgAVAD5QAF5QMNBYAA8UHlJRagAjACtYVFAAflAAb1AAbQBpZgpQAGtYHAk4gF18gC5Qb18AoKJQsIqmqIBfUHawgG4XFyJ9AnFkeDC7FD0UuYWAdSJ6ADEiR0Zil1AxbNyULd39tfai45PQcWvGdsyc1EmToc-QDgQZgDyZ1Qr2BKEmQymMzW8yQi2Wq0YsNglzoewOBAAjEcTm9zqj0TccQ8nhjMaD3uD7t97n9AWCKecIVMkEJhMYUuyCc8CO1Ov5AsFMOEEbNkdyMVFYncTiBQAA5agJAiaAjCe541ASomlJoAGlpmH+FCBlPauoNNJZbI5y21jHJHR8Ap6fVFMM22zRPMxUuJcrCHlAkEMcAIusgoDoCFEFCVbE0ah6diqoFZLBYQlAAHcvSo1MH+d0hSESqS1o7NeDQEMIpAnKAABQAJgArK2AJwASg1YPtvLK+sNxtN53NQ5pQA)

**Related Issues:** 
* https://github.com/microsoft/TypeScript/issues/30082

**Comment by RyanCavanaugh** [comment]
This is the intended behavior. When you create an all-optional type, its assignability actually tightens a bit to only accept source types that have at least one property in common with the target type. `[]` is an array, arrays have a `filter` method, ergo `[]` is assignable to e.g. `{ filter?: unknown }` but not `{ filter1?: unknown }`

**Comment by vasilii-kovalev** [comment]
@RyanCavanaugh, thank you for the explanation. So, am I right saying that there is no way to improve typechecking in this case and restrict type to accept "objects" as "dictionaries" only, not arrays (`isArray` returns `true`) or something like this?

The fields can have "object" type in my examples, because I keep in mind, that there should be dictionaries, but the exact fields and their types will be discussed and defined later, like TODOs as types. And in this case I don't do anything with such fields, but they break the typechecking of the other fields (`objectWithFilter`, for example).
